### ISSUE TITLE
Weird discrepancy in layer stats between this and nodchip learners.

### ISSUE THREAD

**Issue by Sopel97** [glinscott/nnue-pytorch#17]
I started adding some layer statistic to logging, and while doing this I noticed that the values are quite weird. Possible causes that I have no idea about right now would be 1. different quantization. 2. different initialization

For example these are the logs from the pytorch trainer and nodchip trainer (both near the beginning of the training):
```
layer_0:avg_abs_weight tensor(0.0029, device='cuda:0') -- nodchip 0.0136633 WAT
layer_0:avg_abs_bias tensor(0.0391, device='cuda:0') -- nodchip 0.363736 WAT
layer_0:clipped_pct tensor(0.2302, device='cuda:0') -- nodchip 0.4696 WAT
layer_1:avg_abs_weight tensor(0.0323, device='cuda:0') -- nodchip 0.0465173 OK
layer_1:avg_abs_bias tensor(0.0420, device='cuda:0') -- nodchip 0.488686 WAT
layer_2:clipped_pct tensor(0.3352, device='cuda:0') -- nodchip 0.6184 WAT
layer_3:avg_abs_weight tensor(0.1148, device='cuda:0') -- nodchip 0.161746 OK?
layer_3:avg_abs_bias tensor(0.0888, device='cuda:0') -- nodchip 0.520152 WAT
layer_4:clipped_pct tensor(0.2703, device='cuda:0') -- nodchip 0.8391 WAT
layer_5:avg_abs_weight tensor(0.2141, device='cuda:0') -- nodchip 0.442272 WAT
layer_5:avg_abs_bias tensor(0.0650, device='cuda:0') -- nodchip 1.21253 this varies like crazy anyway
Epoch 0:   8%|▌      | 999/12331 [02:51<32:24,  5.83it/s, loss=0.019, v_num=17]C
```
the values in both trainers are reproducible with very small variance over runs (apart from output bias)

Any possible explanations?

**Comment by Sopel97** [comment]
This is cause by very different initialization in the pytorch trainer.
I made a rough sketch of what nodchip initialization would look like:
```
def __init__(self, feature_set=halfkp, lambda_=1.0):
    super(NNUE, self).__init__()
    num_inputs = feature_set.INPUTS
    self.input = nn.Linear(num_inputs, L1)

    # Zero out the weights/biases for the factorized features
    # Weights stored as [256][41024]
    weights = self.input.weight.narrow(1, 0, feature_set.INPUTS - feature_set.FACTOR_INPUTS)
    kMaxActiveDimensions = 32
    kSigma = 0.1 / math.sqrt(kMaxActiveDimensions)
    weights = weights.normal_(0.0, kSigma)
    weights = torch.cat((weights, torch.zeros(L1, feature_set.FACTOR_INPUTS)), dim=1)
    biases = self.input.bias
    biases = biases.clone().fill_(0.5)
    self.input.weight = nn.Parameter(weights)
    self.input.bias = nn.Parameter(biases)

    self.l1 = nn.Linear(2 * L1, L2)
    weights = self.l1.weight
    biases = self.l1.bias
    kSigma = 1.0 / math.sqrt(2 * L1)
    weights = weights.clone().normal_(0.0, kSigma)
    biases = 0.5 - 0.5 * weights.sum(dim=1)
    self.l1.weight = nn.Parameter(weights)
    self.l1.bias = nn.Parameter(biases)

    self.l2 = nn.Linear(L2, L3)
    weights = self.l2.weight
    biases = self.l2.bias
    kSigma = 1.0 / math.sqrt(L2)
    weights = weights.clone().normal_(0.0, kSigma)
    biases = 0.5 - 0.5 * weights.sum(dim=1)
    self.l2.weight = nn.Parameter(weights)
    self.l2.bias = nn.Parameter(biases)

    self.output = nn.Linear(L3, 1)
    weights = self.output.weight
    biases = self.output.bias
    biases = biases.clone().fill_(0.0)
    weights = weights.clone().fill_(0.0)
    self.output.weight = nn.Parameter(weights)
    self.output.bias = nn.Parameter(biases)
```

This produces closer values
```
layer_0:avg_abs_weight tensor(0.0139)
layer_0:avg_abs_bias tensor(0.5000)
layer_0:clipped_pct tensor(0.)
layer_1:avg_abs_weight tensor(0.0352)
layer_1:avg_abs_bias tensor(0.6250)
layer_2:clipped_pct tensor(0.)
layer_3:avg_abs_weight tensor(0.1384)
layer_3:avg_abs_bias tensor(0.5791)
layer_4:clipped_pct tensor(0.)
layer_5:avg_abs_weight tensor(7.1643e-05)
layer_5:avg_abs_bias tensor(0.0002)
```

**Comment by Sopel97** [comment]
Nodchip initialization was shown to be worse in one test. I'll close this issue when there's some data put here.

**Comment by glinscott** [comment]
Taking another look at this for the feature transfomer - see https://github.com/official-stockfish/Stockfish/issues/3274 for details

**Comment by vondele** [comment]
For what it is worth, I took a net that was trained as a restart from master:
https://tests.stockfishchess.org/tests/view/5fce6f3e1ac1691201888179
run50/run0/d09_2B_rescored    : nn-19fa94db0e9e.nnue (-20.98)
and that had no dead spots.
![image](https://user-images.githubusercontent.com/4202567/103383110-ce152c80-4af1-11eb-9d06-210970e850da.png)

**Comment by glinscott** [comment]
I trained two nets, one baseline and one using custom feature transformer initialization (from @Sopel97's nodchip init code, but I only changed the input layer):
```
+    weights = self.input.weight.narrow(1, 0, feature_set.num_features)
+    kMaxActiveDimensions = 32
+    kSigma = 0.1 / math.sqrt(kMaxActiveDimensions)
+    weights = weights.normal_(0.0, kSigma)
+    biases = self.input.bias
+    biases = biases.clone().fill_(0.5)
+    self.input.weight = nn.Parameter(weights)
+    self.input.bias = nn.Parameter(biases)
```

Interestingly, it appears the original was still a fair bit stronger, even with the dead neurons.  Perhaps the other net has more room for growth though - these were both trained on depth 2 data.

Original:
```
Score of master vs version_1-d2-batch8096.nnue: 345 - 206 - 559  [0.563] 1110
Elo difference: 43.7 +/- 14.4, LOS: 100.0 %, DrawRatio: 50.4 %
```

![image](https://user-images.githubusercontent.com/554932/103421543-d309c880-4b51-11eb-9d77-4c9539a1b7d6.png)

New initialization:
```
Score of master vs version_6-init.nnue: 228 - 109 - 292  [0.595] 629
Elo difference: 66.5 +/- 19.9, LOS: 100.0 %, DrawRatio: 46.4 %
```

![image](https://user-images.githubusercontent.com/554932/103421520-b2da0980-4b51-11eb-8cf4-87f86c958001.png)

**Comment by vondele** [comment]
new init looks better to me, at least net seems more like master. 
I wonder if we still need to somehow adjust the loss function, i.e. the lambda doesn't match nodchip yet.

**Comment by ddobbelaere** [comment]
@glinscott Did you train with a flat constant LR throughout the run? I still have a strong suspicion that we are investigating/comparing nets that are not fully trained yet, because of #37.

Note that without dead neurons, the loss landscape becomes more complex, which can be reflected in net performance if it's not yet "maxed out" in training.

**Comment by vondele** [comment]
I have new best net, https://tests.stockfishchess.org/tests/view/5feeebde6019e097de3eead0 and it has even more dead neurons than before (28). That's quite surprising. 

Just wondering... would that mean that we could get the same performance using a feature transformer of smaller size?

**Comment by ddobbelaere** [comment]
That's interesting! Maybe it's an idea to use latest insights (both stepped LR + nodchip init like @glinscott did), but investigate effect of number of neurons by e.g. killing them on purpose (init bias to e.g. -0.5 should do the trick)?

**Comment by glinscott** [comment]
@ddobbelaere I did use a LR step of 0.1 after 75% of training, and it dropped the loss quite a bit.

**Comment by vondele** [comment]
ranger eps = 1e-6:
![image](https://user-images.githubusercontent.com/4202567/103475588-a3a3c780-4dae-11eb-8c66-8bdb583f2cd8.png)
range eps = 1e-8:
![image](https://user-images.githubusercontent.com/4202567/103475597-ba4a1e80-4dae-11eb-9e23-eeb5b3bdb67f.png)

**Comment by Sopel97** [comment]
I think we've determined a good initialization scheme. Nodchip's doesn't seem to work well.

As for the epsilon, some results here https://docs.google.com/document/d/1gTlrr02qSNKiXNZ_SuO4-RjK4MXBiFlLE6jvNqqMkAY/edit#heading=h.93s29zcljbj2
